// @netlify/database/dist/main.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/main.ts
var main_exports = {};
__export(main_exports, {
  MissingDatabaseConnectionError: () => MissingDatabaseConnectionError,
  getConnectionString: () => getConnectionString,
  getDatabase: () => getDatabase
});
module.exports = __toCommonJS(main_exports);
var import_serverless = require("@neondatabase/serverless");
var import_ws = __toESM(require("ws"), 1);
var import_pg = __toESM(require("pg"), 1);
var import_neon_http = require("waddler/neon-http");
var import_node_postgres = require("waddler/node-postgres");
var import_runtime_utils = require("@netlify/runtime-utils");

// src/environment.ts
var MissingDatabaseConnectionError = class extends Error {
  constructor() {
    super(
      "The environment has not been configured to use Netlify Database. You must supply the `connectionString` option when calling `getDatabase()`. See https://ntl.fyi/database-environment for details."
    );
    this.name = "MissingDatabaseConnectionError";
  }
};

// src/main.ts
function getConnectionString() {
  const env = (0, import_runtime_utils.getEnvironment)();
  const connectionString = env.get("NETLIFY_DB_URL");
  if (!connectionString) {
    throw new MissingDatabaseConnectionError();
  }
  return connectionString;
}
function ensureNeonWebSocket() {
  if (!import_serverless.neonConfig.webSocketConstructor && typeof WebSocket === "undefined") {
    import_serverless.neonConfig.webSocketConstructor = import_ws.default;
  }
}
function createRefreshingHttpClient(readConnectionString) {
  let cachedConnectionString;
  let cachedClient;
  const getClient = () => {
    const connectionString = readConnectionString();
    if (connectionString !== cachedConnectionString || cachedClient === void 0) {
      cachedConnectionString = connectionString;
      cachedClient = (0, import_serverless.neon)(connectionString);
    }
    return cachedClient;
  };
  getClient();
  const target = (() => void 0);
  return new Proxy(target, {
    apply: (_target, _thisArg, args) => getClient()(...args),
    get: (_target, prop, receiver) => {
      const value = Reflect.get(getClient(), prop, receiver);
      if (typeof value !== "function") {
        return value;
      }
      return (...args) => {
        const fn = Reflect.get(getClient(), prop, receiver);
        return fn.apply(getClient(), args);
      };
    }
  });
}
function getDatabase(options = {}) {
  const env = (0, import_runtime_utils.getEnvironment)();
  const { connectionString: override } = options;
  const readConnectionString = () => {
    const connectionString2 = override ?? env.get("NETLIFY_DB_URL");
    if (!connectionString2) {
      throw new MissingDatabaseConnectionError();
    }
    return connectionString2;
  };
  const connectionString = readConnectionString();
  const driver = env.get("NETLIFY_DB_DRIVER");
  if (driver === "serverless") {
    ensureNeonWebSocket();
    const httpClient = createRefreshingHttpClient(readConnectionString);
    const pool2 = new import_serverless.Pool({ connectionString });
    return {
      driver: "serverless",
      sql: (0, import_neon_http.waddler)({ client: httpClient }),
      pool: pool2,
      httpClient,
      get connectionString() {
        return readConnectionString();
      }
    };
  }
  const pool = new import_pg.default.Pool({ connectionString });
  return {
    driver: "server",
    sql: (0, import_node_postgres.waddler)({ client: pool }),
    pool,
    get connectionString() {
      return readConnectionString();
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MissingDatabaseConnectionError,
  getConnectionString,
  getDatabase
});
