// @netlify/database/dist/main.js
// src/main.ts
import { neon, neonConfig, Pool as NeonPool } from "@neondatabase/serverless";
import ws from "ws";
import pg from "pg";
import { waddler as waddlerNeonHttp } from "waddler/neon-http";
import { waddler as waddlerNodePostgres } from "waddler/node-postgres";
import { getEnvironment } from "@netlify/runtime-utils";

// src/environment.ts
var MissingDatabaseConnectionError = class extends Error {
  constructor() {
    super(
      "The environment has not been configured to use Netlify Database. You must supply the `connectionString` option when calling `getDatabase()`. See https://ntl.fyi/database-environment for details."
    );
    this.name = "MissingDatabaseConnectionError";
  }
};

// src/main.ts
function getConnectionString() {
  const env = getEnvironment();
  const connectionString = env.get("NETLIFY_DB_URL");
  if (!connectionString) {
    throw new MissingDatabaseConnectionError();
  }
  return connectionString;
}
function ensureNeonWebSocket() {
  if (!neonConfig.webSocketConstructor && typeof WebSocket === "undefined") {
    neonConfig.webSocketConstructor = ws;
  }
}
function createRefreshingHttpClient(readConnectionString) {
  let cachedConnectionString;
  let cachedClient;
  const getClient = () => {
    const connectionString = readConnectionString();
    if (connectionString !== cachedConnectionString || cachedClient === void 0) {
      cachedConnectionString = connectionString;
      cachedClient = neon(connectionString);
    }
    return cachedClient;
  };
  getClient();
  const target = (() => void 0);
  return new Proxy(target, {
    apply: (_target, _thisArg, args) => getClient()(...args),
    get: (_target, prop, receiver) => {
      const value = Reflect.get(getClient(), prop, receiver);
      if (typeof value !== "function") {
        return value;
      }
      return (...args) => {
        const fn = Reflect.get(getClient(), prop, receiver);
        return fn.apply(getClient(), args);
      };
    }
  });
}
function getDatabase(options = {}) {
  const env = getEnvironment();
  const { connectionString: override } = options;
  const readConnectionString = () => {
    const connectionString2 = override ?? env.get("NETLIFY_DB_URL");
    if (!connectionString2) {
      throw new MissingDatabaseConnectionError();
    }
    return connectionString2;
  };
  const connectionString = readConnectionString();
  const driver = env.get("NETLIFY_DB_DRIVER");
  if (driver === "serverless") {
    ensureNeonWebSocket();
    const httpClient = createRefreshingHttpClient(readConnectionString);
    const pool2 = new NeonPool({ connectionString });
    return {
      driver: "serverless",
      sql: waddlerNeonHttp({ client: httpClient }),
      pool: pool2,
      httpClient,
      get connectionString() {
        return readConnectionString();
      }
    };
  }
  const pool = new pg.Pool({ connectionString });
  return {
    driver: "server",
    sql: waddlerNodePostgres({ client: pool }),
    pool,
    get connectionString() {
      return readConnectionString();
    }
  };
}
export {
  MissingDatabaseConnectionError,
  getConnectionString,
  getDatabase
};
